src/iterator.rs

```diff
@@ -0,0 +1,54 @@
+use freight_vm::{error::FreightError, execution_engine::ExecutionEngine};
+use std::cell::RefCell;
+use std::fmt::Debug;
+use std::rc::Rc;
+
+use crate::fender_value::FenderValue;
+use crate::{fender_reference::FenderReference, type_sys::type_system::FenderTypeSystem};
+
+pub type NativeClosure =
+    Rc<dyn Fn(&mut ExecutionEngine<FenderTypeSystem>) -> Result<FenderReference, FreightError>>;
+
+#[derive(Clone)]
+pub struct FenderIterator {
+    pub next: NativeClosure,
+    pub has_next: NativeClosure,
```

Comment:
instead of or in addition to a `has_next` I think we should use a return type and actaully give iterators an internal state that can be accessed on its own instead of just moving a rust iterator inside of a closure

Reply:
We had discussed this already, and I had said I was going to do it this way because I couldn't see any good way to do it with a single function. This design is the most flexible and lightweight I could think of. If we have another struct that's wrapping every returned item, it requires a heap allocation for every item returned. I'm not a fan of that. We'd also just have a new type whose only purpose is to wrap other types, and I don't like that either.

If we give iterators an internal state, then we run into the issue of it needing to always been a FenderReference. We can't parameterize the type. Keeping it as just a FenderReference really limits what we can meaningfully (and easily) use as state.

I really prefer doing it this way because it doesn't require users to unwrap any other values. I've been thinking about that design and I just can't bring myself to like it because it feels less explicit. That design works really well for rust *because* it is a statically-typed language with a powerful type system. Because fender is not, I much prefer hasNext and next over just a single next function that returns a wrapped value. hasNext also allows you to preview whether the iterator has another item without retrieving it.